### ISSUE TITLE
Include ParaLLEl-RDP as a git subtree

### ISSUE THREAD

**Issue by Screwtapello** [ares-emulator/ares#5]
ares' N64 core depends on https://github.com/Themaister/parallel-rdp-standalone for RDP emulation. Currently, ares includes [a script](https://github.com/higan-emu/ares/blob/5c697600f7d997cd456a596df1f1f3ef2f7dbdb6/ares/n64/vulkan/GNUmakefile) to download the latest version of that code in the right place in ares' source tree, but now that ares itself is stored in git, we can do better.

- `git submodule` is the official tool for putting git repos inside other git repos. The upside is that there is a standard "update all submodules" command people are already likely to know; the downside is that submodules aren't checked out by default and the failure modes of not-quite-right submodules can be confusing.
- `git subtree` is an alternative tool. The upside is that the included repository's files are stored directly in the ares repo so that they can't get lost or misplaced; the downside is that there's no way to automatically update all subtrees.

For example, higan uses git subtree for both the external libco repository, and for hunterk's shader repository. I put an [update-subtrees.sh](https://github.com/higan-emu/higan/blob/354efff9374166824f4f4b31708f270459624c67/update-subtrees.sh) script in the root of the repository to make it easy to update subtrees in future, but since that's a project-specific convention it would need to be documented somewhere.

Of course, we could also just keep the status quo, and require people to run the script to obtain ParaLLEl-RDP before building the N64 core.

**Comment by LukeUsher** [comment]
We should post some build instructions somewhere, but a full build, including updating ParaLLEl is as simple as running one command: 

`make -C ares/n64/vulkan sync-upstream && make -C lucia`

So it's not currently a major blocker. I updated the CI scripts to do this automatically.

**Comment by qwertymodo** [comment]
The best/simplest explanation I've seen for choosing between subtree vs submodule comes down to the workflow.  If the dependency doesn't update often, and you want to control exactly when to change the dependency version that you're using in the main repo (e.g. the dependency is developed independently and you only want to update it when stable releases become available), then a submodule is the better choice.  If you're developing both in tandem and want to keep both constantly up-to-date, then a subtree is the easier way to achieve that.  Either way, I think using the native git tooling for this is a good idea.  If we're worried about external dependencies and the possibility of them disappearing, or other problems due to them not being under our control, we could always fork the repo here and set our own fork as the linked dependency.

**Comment by LukeUsher** [comment]
I'm leaning towards a submodule, purely because we are not likely to be contributing much to parallel-rdp itself,  and we can potentially automate keeping parallel-rdp current with standard git procedure.

The issues that occur with submodules shouldn't be a problem here, because we will never be modifying the tree.

**Comment by Screwtapello** [comment]
The issues I've had (and co-workers have had, etc.) with submodules are that `git clone` doesn't fetch them unless you add `--recurse-submodules`, `git checkout` doesn't update them without `--recurse-submodules`, etc. And once you *do* wind up in a scenario where some submodules are missing or out-of-date, it's not clear how to fix that: `git submodule init` is required but doesn't appear to fix anything, `git submodule update` *keeps* submodules out-of-date instead of updating them to the latest version, etc. I predict that if ares includes ParaLLEl-RDP as a submodule, there will be predictable and regular tech-support issues filed about it being missing or out-of-date.

Additionally, if somebody wanted to download a source archive of ares for some reason (to build on an offline computer or similar) it's a lot more work to include submodule data - you'd need to make a special "source archive" release artifact; tools like `git archive` and GitHub's "download source" button would be broken.

I'm not saying that git subtree is perfect, and the issues I mentioned above might not be impontant or significant for ares in practice. However, I do believe `git subtree` is the easiest way to vendor one git repo into another, and the most reliable since even if you don't know the subtree exists, it Just Works.

**Comment by LukeUsher** [comment]
those are some good points.

With Cxbx-R we have a script in the repo that does `git submodule update --init --recursive` which solved everything in a single command (inits the submodules, syncs with the latest commit linked to the repo).

I think there is some confusion between what `git submodule update` does: it is not intended to update the submodule to the latest upstream version; it is intended to update your local copy of the submodule to the version specified in the git repository you're in, to update, you have to cd into the submodule and pull from upstream, and then commit that change.

Submodules are the only solution that will keep 'zip' downloads of the source code working, however, so maybe that is the best solution. I'm just totally unfamiliar with submodules and would likely need a crash course on how to use/update them.

**Comment by Screwtapello** [comment]
From a fresh ares checkout, adding ParaLLEl-RDP as a subtree is:

```
git subtree add \
    --prefix=ares/n64/vulkan/parallel-rdp \
    https://github.com/ares-emu/parallel-rdp.git \
    HEAD \
    --squash
```

- `--prefix` says where to put the subtree contents
- `https://...` is the repository to fetch from
- `HEAD` is the ref to fetch from that repository
- `--squash` means "squash the entire history of the repo into a single commit"

The alternative to `--squash` is to include *all* the commits from the upstream repo into the commit history, which can be useful for tracking down regressions, but since the ParaLLEl-RDP repo we pull from is automatically generated, the commit messages aren't very useful and would just add a lot of noise in this instance.

This command adds two commits to the current branch:

```gitlog
*   commit ea983c4ee854fc4b8811f0f5a51a5305d25a62d3 (HEAD -> parallel-rdp-subtree)
|\  Merge: 7adfece67 3c4528049
| | Author: Tim Allen <screwtape@froup.com>
| | Date:   Fri Aug 6 10:39:22 2021 +1000
| | 
| |     Merge commit '3c45280490bacf28c4e1a81b93c835485d6a4d77' as 'ares/n64/vulkan/parallel-rdp'
| | 
| * commit 3c45280490bacf28c4e1a81b93c835485d6a4d77
|   Author: Tim Allen <screwtape@froup.com>
|   Date:   Fri Aug 6 10:39:22 2021 +1000
|   
|       Squashed 'ares/n64/vulkan/parallel-rdp/' content from commit 0097af2f4
|       
|       git-subtree-dir: ares/n64/vulkan/parallel-rdp
|       git-subtree-split: 0097af2f4d1f880d403c150f5fc4d55d825cd799
| 
```

Commit 3c45280 is the squashed history of the repo; if you `git show` it you'll see all the original files in their original paths. The commit message includes some git subtree metadata including the original commit ID, so that it can efficiently grab changes in future.

Commit ea983c4 is the merge commit that rewrites the paths of all the files so they appear in the correct location in the filesystem.

Once a subtree has been added, updating it looks exactly the same, but with "pull" instead of "add":

```
git subtree pull \
    --prefix=ares/n64/vulkan/parallel-rdp \
    https://github.com/ares-emu/parallel-rdp.git \
    HEAD \
    --squash
```

If I do it immediately, there hasn't been any updates to the repo yet, so I get:

```
Subtree is already at commit 0097af2f4d1f880d403c150f5fc4d55d825cd799.
```

If I decide to pull Themaister's upstream instead of Near's fork (which just has a change to quiet a compiler warning), I get new commits added on top of the ones above:

```gitlog
*   commit 7ccb8820c24eeee307cbafb1ce3979f5b9722e0b (HEAD -> parallel-rdp-subtree)
|\  Merge: ea983c4ee 24f4651e0
| | Author: Tim Allen <screwtape@froup.com>
| | Date:   Fri Aug 6 10:57:23 2021 +1000
| | 
| |     Merge commit '24f4651e015b047c91d8220ed9ee003db88c3ca5' into parallel-rdp-subtree
| | 
| * commit 24f4651e015b047c91d8220ed9ee003db88c3ca5
| | Author: Tim Allen <screwtape@froup.com>
| | Date:   Fri Aug 6 10:57:23 2021 +1000
| | 
| |     Squashed 'ares/n64/vulkan/parallel-rdp/' changes from 0097af2f4..464ab61fc
| |     
| |     REVERT: 0097af2f4 Update vulkan_headers.hpp
| |     
| |     git-subtree-dir: ares/n64/vulkan/parallel-rdp
| |     git-subtree-split: 464ab61fce58b27a0e0ed9f0e366aa1113694965
| | 
* | commit ea983c4ee854fc4b8811f0f5a51a5305d25a62d3
|\| Merge: 7adfece67 3c4528049
| | (...as above...)
```

There's a new squashed commit that mentions the commit in the upstream repo, and records that Near's quiet-warnings commit was reverted, and then another merge commit that folds those changes back into the appropriate subdirectory of the main branch.

I usually stick all the pull commands in [an update-subtrees.sh script](https://github.com/higan-emu/higan/blob/354efff9374166824f4f4b31708f270459624c67/update-subtrees.sh) so anyone can easily update the dependencies; even if they're not on a platform with a POSIX shell, it's easy to see what commands you need to run to update things.

If I were making a PR, the next step would be to add a commit that removes the Makefile Near used to sync with upstream, and another to add an `update-subtrees.sh` script, and that would be it.

**Comment by LukeUsher** [comment]
Excellent!

**Comment by LukeUsher** [comment]
fixed in c2fcacb331f41aa57337a1638b8a29c71941ea9f
